### PR TITLE
Add DYNAMIC_ROUTER setting

### PR DESCRIPTION
**Pull request by iksaif**
The idea is that you want your consistent hashing ring to
adapt itself when destinations are down. This change will
remove destinations from the router when down and add them
back when up.

An additional buffer is created to store points when no
destination is up, the receiving socket is also paused in this
case to apply some back-pressure.

When a node is removed from the router its points are re-injected
to carbon to go through the new route.

### CODE REVIEW

**Comment by iksaif** [comment]
Can I haz review ?

**Comment by erebe** [comment]
looking

File: lib/carbon/client.py [hunk 1]
```diff
@@ -368,11 +425,45 @@ def clientProtocol(self):
...
 
+class FakeClientFactory(object):
+  """Fake client factory that buffers points
+
+  This is used when all the destinations are down and before we
+  pause the reception of metrics to avoid loosing points.
+  """
+
+  def __init__(self):
+    self.queue = deque()
```
Comment (erebe) [inline lib/carbon/client.py]: Can you make it bounded, even if it is a big number.

File: lib/carbon/client.py [hunk 2]
```diff
@@ -306,20 +315,68 @@ def sendHighPriorityDatapoint(self, metric, datapoint):
...
+      log.clients("Adding client %s:%d:%s to router" % destination)
+      self.router.addDestination(destination)
+      state.events.resumeReceivingMetrics()
+
+  def destinationDown(self, destination):
+    # Only blacklist the destination if we tried a lot.
+    log.clients("Destination is down: %s:%d:%s (%d/%d)" % (
+        destination[0], destination[1], destination[2], self.retries,
+        settings.DYNAMIC_ROUTER_MAX_RETRIES))
+    if self.retries < settings.DYNAMIC_ROUTER_MAX_RETRIES:
```
Comment (erebe) [inline lib/carbon/client.py]: where do you increment retries ?

Comment (iksaif) [reply lib/carbon/client.py]: ReconnectingClientFactory does that

**Review by iksaif — COMMENTED** [review]
I added comment in the code to explain why it looks like that

**Comment by erebe** [comment]
:+1:

**Comment by iksaif** [comment]
Will push early next week because it needs to be monitored closely

**Pull request merged**